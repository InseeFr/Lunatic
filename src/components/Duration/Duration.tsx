import './Duration.scss';
import type { LunaticComponentProps } from '../type';
import { useState } from 'react';
import { getDurationFromValue } from './getDurationFromValue';
import { clampDuration, labelByUnit, propsByUnit } from './durationUtils';
import { formatDuration } from './formatDuration';
import classnames from 'classnames';
import { objectKeys } from '../../utils/object';
import { customizedComponent } from '../shared/HOC/customizedComponent';
import type { LunaticError } from '../../use-lunatic/type';
import {
	ComponentErrors,
	getComponentErrors,
} from '../shared/ComponentErrors/ComponentErrors';
import { Declarations } from '../shared/Declarations/Declarations';

export function Duration({
	handleChange,
	response,
	errors,
	...props
}: LunaticComponentProps<'Duration'>) {
	return (
		<CustomDuration
			{...props}
			onChange={(v) => handleChange(response, v)}
			errors={getComponentErrors(errors, props.id)}
		/>
	);
}

type CustomProps = Omit<
	LunaticComponentProps<'Duration'>,
	'response' | 'handleChange' | 'errors'
> & {
	onChange: (v: string | null) => void;
	errors?: LunaticError[];
};

export const CustomDuration = customizedComponent<CustomProps>(
	'Duration',
	(props) => {
		const { errors, value, label, format, onChange, declarations, id } = props;
		// We need to keep an internal state since one field can be empty (null value in duration)
		// but we still send "0" and we don't want the field to display "0"
		const [duration, setDuration] = useState(
			getDurationFromValue(value, format)
		);

		// Generate handler for a specific unit field (year, month...)
		const changeHandler =
			(unit: 'hours' | 'minutes' | 'months' | 'years') =>
			(e: {
				// CheckValidity function is used to apply constraints to a field Ex: (min, max)
				target: { valueAsNumber: number; checkValidity: () => boolean };
			}) => {
				if (!e.target.checkValidity()) {
					return;
				}
				const fieldValue = clampDuration(
					Number.isNaN(e.target.valueAsNumber) ? null : e.target.valueAsNumber,
					unit
				);
				const newDuration = { ...duration, [unit]: fieldValue };
				onChange(formatDuration(newDuration));
				setDuration(newDuration);
			};

		return (
			<fieldset className={classnames('lunatic-input')}>
				{label && <legend>{label}</legend>}
				<Declarations
					type="AFTER_QUESTION_TEXT"
					declarations={declarations}
					id={id}
				/>
				<div className="duration-fields">
					{objectKeys(duration).map((unit) => (
						<div className="duration-field" key={unit}>
							<label htmlFor={`${unit}Input`} className="input-label">
								{labelByUnit[unit]}
							</label>
							<input
								id={`${unit}Input`}
								type="number"
								value={duration[unit] ?? ''}
								className="input-field"
								onChange={changeHandler(unit)}
								{...propsByUnit[unit]}
							/>
						</div>
					))}
				</div>
				<ComponentErrors errors={errors} />
			</fieldset>
		);
	}
);
