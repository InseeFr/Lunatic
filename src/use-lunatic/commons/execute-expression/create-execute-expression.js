import executeExpression from './execute-expression';
import getSafetyExpression from './get-safety-expression';
import getExpressionVariables from './get-expressions-variables';
import createMemoizer from './create-memoizer';
import createRefreshCalculated from './create-refresh-calculated';
import getVtlCompatibleValue from 'utils/vtl';
import { VTL, VTL_MD, X_AXIS, Y_AXIS } from 'utils/constants';

function validateExpression(expObject) {
	if (typeof expObject === 'object') {
		const { type } = expObject;
		if (type === VTL || type === VTL_MD) {
			return expObject;
		}
	}
	throw new Error(`Non-VTL compatible expression : ${expObject}`);
}

function createBindings(variables) {
	return Object.entries(variables).reduce(function (
		bindings,
		[name, { value }]
	) {
		return { ...bindings, [name]: value };
	},
	{});
}

/**
 *
 * @param {*} variables
 * @returns
 */
function createExecuteExpression(variables, features) {
	// on aimerait map d'expression, avec les bindings
	const bindings = createBindings(variables);
	const tokensMap = new Map();
	const collectedUpdated = new Map();
	const [memoize, getMemoizedValue] = createMemoizer();
	const [refreshCalculated, setToRefreshCalculated] = createRefreshCalculated({
		variables,
		execute,
		bindings,
	});

	/**
	 *
	 * @param {*} name
	 */
	function pushToLazy(name) {
		const { CalculatedLinked = [] } = variables[name];

		CalculatedLinked.forEach(function (variable) {
			const { name } = variable;
			setToRefreshCalculated(name, variable);
		});
	}

	/**
	 *
	 * @param {*} name
	 * @param {*} value
	 */
	function updateBindings(name, value) {
		// update des bindings
		if (name in bindings) {
			bindings[name] = value;
			collectedUpdated.set(name, []);
		}
		pushToLazy(name);
	}

	function getVariablesAndCach(expression) {
		if (tokensMap.has(expression)) {
			return tokensMap.get(expression);
		}
		const tokens = getExpressionVariables(expression, variables);
		tokensMap.set(expression, tokens);
		return tokens;
	}

	/**/
	function collecteVariables(dependencies) {
		if (Array.isArray(dependencies)) {
			return dependencies.reduce(function (map, name) {
				if (name in variables) {
					const data = variables[name];
					const { variable, type } = data;
					if (!(name in map)) {
						if (type === 'CALCULATED') {
							const { expression } = variable;
							const subDependencies = getVariablesAndCach(expression);

							return {
								...map,
								[name]: { ...variable },
								...collecteVariables(subDependencies),
							};
						}

						return { ...map, [name]: { ...variable } };
					}
				} else {
					throw new Error(`Unknown variable ${name}`);
				}
				return map;
			}, {});
		}
		return {};
	}

	function resolveUseContext(name, { iteration, linksIterations }) {
		const value = bindings[name];
		if ([X_AXIS, Y_AXIS].includes(name) && linksIterations !== undefined) {
			pushToLazy(name);
			const [x, y] = linksIterations;
			if (Array.isArray(value) && x < value.length) {
				return value[name === X_AXIS ? x : y];
			}
			return null;
		}
		if (iteration !== undefined && Array.isArray(value)) {
			pushToLazy(name);
			if (iteration < value.length) {
				return value[iteration];
			}
			return null;
		}
		return getVtlCompatibleValue(value);
	}

	function fillVariablesValues(map, { iteration, linksIterations }) {
		return Object.entries(map).reduce(function (sub, [name, _]) {
			return {
				...sub,
				[name]: resolveUseContext(name, { iteration, linksIterations }),
			};
		}, {});
	}

	/**
	 *
	 * @param {*} vtlObject
	 * @param {*} args
	 * @returns
	 */
	function execute(expObject, args = {}) {
		const { value: expression, type } = validateExpression(
			getSafetyExpression(expObject)
		);
		const { iteration, linksIterations, logging } = args;
		const bindingDependencies = getVariablesAndCach(expression);

		function loggingDefault(_, bindings, e) {
			if (process.env.NODE_ENV === 'development') {
				console.warn(`VTL error :  ${expression}`, { ...args }, { bindings });
				console.warn(e);
			}
		}
		const vtlBindings = refreshCalculated(
			fillVariablesValues(collecteVariables(bindingDependencies), {
				iteration,
				linksIterations,
			}),
			{ rootExpression: expression, iteration, linksIterations }
		);

		const memoized = getMemoizedValue(expression, vtlBindings);
		if (memoized === undefined) {
			const result = executeExpression(
				vtlBindings,
				expression,
				type,
				features,
				logging || loggingDefault
			);
			memoize(expression, vtlBindings, result);

			return result;
		}
		return memoized;
	}

	return [execute, updateBindings];
}

export default createExecuteExpression;
