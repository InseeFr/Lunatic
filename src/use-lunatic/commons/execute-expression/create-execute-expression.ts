import executeExpression from './execute-expression';
import getSafetyExpression from './get-safety-expression';
import getExpressionVariables from './get-expressions-variables';
import createMemoizer from './create-memoizer';
import createRefreshCalculated from './create-refresh-calculated';
import { getVTLCompatibleValue } from '../../../utils/vtl';
import { VTL, VTL_MD, X_AXIS, Y_AXIS } from '../../../utils/constants';
import type { LunaticExpression, LunaticState } from '../../type';

export type ExpressionLogger = (
	expression: string | LunaticExpression,
	bindings: { [variableName: string]: unknown },
	e: unknown
) => void;
type Bindings = { [variableName: string]: unknown };

/**
 * Check the shape of the expression and convert it into an expression
 */
function validateExpression(expObject: unknown): LunaticExpression {
	if (typeof expObject === 'object' && expObject && 'type' in expObject) {
		const { type } = expObject;
		if (type === VTL || type === VTL_MD) {
			return expObject as LunaticExpression;
		}
	}
	if (typeof expObject === 'string') return { value: expObject, type: VTL };

	console.warn(`Non-VTL compatible expression`);
	console.warn(expObject);
	throw new Error(`Uncompatible VTL typeof element : ${typeof expObject}`);
}

/**
 * Extract values for every variables
 */
function createBindings(variables: LunaticState['variables']): Bindings {
	return Object.entries(variables).reduce(function (
		bindings,
		[name, { value }]
	) {
		return { ...bindings, [name]: value };
	},
	{});
}

/**
 * Generates methods to interact with variables
 */
function createExecuteExpression(
	variables: LunaticState['variables'],
	features: string[]
) {
	// on aimerait map d'expression, avec les bindings
	const bindings = createBindings(variables);
	const tokensMap = new Map<LunaticExpression | string, string[]>();
	const [memoize, getMemoizedValue] = createMemoizer();
	const [refreshCalculated, setToRefreshCalculated] = createRefreshCalculated({
		variables,
		execute,
		bindings,
	});

	/**
	 * Mark all linked variables to be refreshed
	 */
	function pushToLazy(name: string) {
		if (name in variables) {
			const { CalculatedLinked = [] } = variables[name];
			CalculatedLinked.forEach(function (variable) {
				const { name } = variable;
				setToRefreshCalculated(name, variable);
			});
		} else {
			console.warn(`${name} is not identified as varaible!`);
		}
	}

	/**
	 * Update the value of a variable
	 */
	function updateBindings(name: string, value: unknown) {
		if (name in bindings) {
			bindings[name] = value;
		}
		pushToLazy(name);
	}

	/**
	 * Extract variables from an expression
	 */
	function getVariablesAndCatch(expression: string | LunaticExpression) {
		// If we have an expression we can't have tokens
		if (typeof expression !== 'string') {
			return [];
		}
		if (tokensMap.has(expression)) {
			return tokensMap.get(expression);
		}
		const tokens = getExpressionVariables(expression, variables);
		tokensMap.set(expression, tokens);
		return tokens;
	}

	/**
	 * Retrieve variable affected by the dependencies
	 */
	function collecteVariables(dependencies: unknown): {
		[name: string]: LunaticState['variables'];
	} {
		if (Array.isArray(dependencies)) {
			return dependencies.reduce(function (map, name) {
				if (name in variables) {
					const data = variables[name];
					const { variable } = data;
					if (!(name in map)) {
						if (variable.variableType === 'CALCULATED') {
							const { expression } = variable;
							const subDependencies = getVariablesAndCatch(expression);

							return {
								...map,
								[name]: { ...variable },
								...collecteVariables(subDependencies),
							};
						}

						return { ...map, [name]: { ...variable } };
					}
				} else {
					throw new Error(`Unknown variable ${name}`);
				}
				return map;
			}, {});
		}
		return {};
	}

	function resolveUseContext(
		name: string,
		{
			iteration,
			linksIterations,
		}: { iteration?: number; linksIterations?: number[] }
	) {
		const value = bindings[name];

		if ([X_AXIS, Y_AXIS].includes(name) && linksIterations !== undefined) {
			pushToLazy(name);
			const [x, y] = linksIterations;
			if (Array.isArray(value) && x < value.length) {
				return value[name === X_AXIS ? x : y];
			}
			return null;
		}
		if (iteration !== undefined && Array.isArray(value)) {
			pushToLazy(name);
			if (iteration < value.length) {
				return value[iteration];
			}
			return null;
		}
		if (linksIterations !== undefined) {
			const [x, y] = linksIterations;
			if (Array.isArray(value) && x < value.length) {
				const sub = value[x];
				if (Array.isArray(sub) && y < sub.length) {
					return sub[y];
				}
			}

			return null;
		}
		return getVTLCompatibleValue(value);
	}

	function fillVariablesValues(
		map: { [variableName: string]: unknown },
		{
			iteration,
			linksIterations,
		}: { iteration?: number; linksIterations?: number[] }
	) {
		return Object.entries(map).reduce(function (sub, [name, _]) {
			return {
				...sub,
				[name]: resolveUseContext(name, { iteration, linksIterations }),
			};
		}, {});
	}

	function execute<T = unknown>(
		expObject: unknown, // VtlExpression
		args: Parameters<LunaticState['executeExpression']>[1] = {}
	): T {
		const { value: expression, type } = validateExpression(
			getSafetyExpression(expObject)
		);
		const { iteration, linksIterations, logging } = args;
		const bindingDependencies = getVariablesAndCatch(expression);

		function loggingDefault(_: unknown, bindings: Bindings, e: unknown) {
			if (process.env.NODE_ENV === 'development') {
				console.warn(`VTL error :  ${expression}`, { ...args }, { bindings });
				console.warn(e);
			}
		}

		const vtlBindings = refreshCalculated(
			fillVariablesValues(collecteVariables(bindingDependencies), {
				iteration,
				linksIterations,
			}),
			{ rootExpression: expression, iteration, linksIterations }
		);
		// Add index has a specific variable
		vtlBindings['GLOBAL_ITERATION_INDEX'] = (
			(args?.iteration ?? 0) + 1
		).toString();
		const memoized = getMemoizedValue(expression, vtlBindings);
		if (memoized === undefined) {
			const result = executeExpression(
				vtlBindings,
				expression,
				type,
				features,
				logging || loggingDefault
			);
			memoize(expression, vtlBindings, result);

			return result as T; // We need to force this type to match with the signature
		}
		return memoized;
	}

	return [execute, updateBindings] as const;
}

export default createExecuteExpression;
