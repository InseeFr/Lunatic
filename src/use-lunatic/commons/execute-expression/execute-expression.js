import React from 'react';
import { interpret } from '@inseefr/trevas';
import MdLabel from '../../../components/commons/components/md-label';
import getVtlCompatibleValue from '../../../utils/vtl';
import { VTL, MD } from '../../../utils/constants';

function isDataSet(result) {
	return result && typeof result === 'object' && result.dataPoints;
}

function extractDataSetResult(dataSet) {
	const { dataPoints } = dataSet;
	if (dataPoints) {
		const { result } = dataPoints;
		return result;
	}
	return undefined;
}

export function executeVtlExpression(expression, vtlBindings) {
	const legalVtlBindings = Object.entries(vtlBindings).reduce(
		(acc, [k, v]) => ({ ...acc, [k]: getVtlCompatibleValue(v) }),
		{}
	);
	const result = interpret(expression, legalVtlBindings);
	if (isDataSet(result)) {
		return extractDataSetResult(result);
	}
	if (result === null) {
		return null;
	}
	if (typeof result === 'object') {
		return expression;
	}

	return result;
}

function executeVTL(expression, vtlBindings, logging) {
	if (expression) {
		try {
			return executeVtlExpression(expression, vtlBindings);
		} catch (e) {
			logging(expression, vtlBindings, e);
			return expression;
		}
	}
	return '';
}

function loggingDefault(expression, bindings, e) {
	if (process.env.NODE_ENV === 'development') {
		console.warn(`VTL error :  ${expression}`, bindings);
		console.warn(e);
	}
}

function executeExpression(
	vtlBindings,
	expression,
	type,
	features /* VTL, MD */,
	logging = loggingDefault
) {
	if (features.includes(VTL) && type.includes(VTL)) {
		const vtl = executeVTL(expression, vtlBindings, logging);
		if (features.includes(MD) && type.includes(MD))
			return <MdLabel expression={vtl} />;
		return vtl;
	}
	return expression;
}

export default executeExpression;
