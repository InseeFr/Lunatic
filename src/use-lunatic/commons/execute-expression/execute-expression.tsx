import { type ReactNode } from 'react';
import { interpret } from '@inseefr/trevas';
import MdLabel from '../../../components/commons/components/md-label';
import getVtlCompatibleValue from '../../../utils/vtl';
import { MD, VTL } from '../../../utils/constants';
import type { VTLBindings } from '../../type';

type DataSet = { dataPoints: { result: unknown } };

function isDataSet(result: unknown): result is DataSet {
	return (
		typeof result === 'object' &&
		result !== null &&
		'dataPoints' in result &&
		result.dataPoints !== null &&
		typeof result.dataPoints === 'object' &&
		'result' in result.dataPoints
	);
}

function extractDataSetResult(dataSet: DataSet) {
	const { dataPoints } = dataSet;
	if (dataPoints) {
		const { result } = dataPoints;
		return result;
	}
	return undefined;
}

/**
 * Run a VTL expression and return the result
 */
export function executeVtlExpression<T = unknown>(
	expression: string,
	vtlBindings: { [variableName: string]: unknown }
) {
	// Make sur all values can be handled by VTL
	const legalVtlBindings = Object.entries(vtlBindings).reduce(
		(acc, [k, v]) => ({ ...acc, [k]: getVtlCompatibleValue(v) }),
		{}
	);
	const result = interpret(expression, legalVtlBindings);
	if (isDataSet(result)) {
		return extractDataSetResult(result);
	}
	if (result === null) {
		return null;
	}
	if (typeof result === 'object') {
		return expression;
	}

	return result;
}

function executeVTL(
	expression: string,
	vtlBindings: VTLBindings,
	logging: typeof loggingDefault
) {
	if (expression) {
		try {
			return executeVtlExpression(expression, vtlBindings);
		} catch (e) {
			logging(expression, vtlBindings, e);
			return expression;
		}
	}
	return '';
}

function loggingDefault(expression: string, bindings: VTLBindings, e: unknown) {
	if (process.env.NODE_ENV === 'development') {
		console.warn(`VTL error :  ${expression}`, bindings);
		console.warn(e);
	}
}

function executeExpression(
	vtlBindings: VTLBindings,
	expression: string,
	type: 'VTL' | 'VTL|MD',
	features: string[],
	logging = loggingDefault
): ReactNode {
	if (features.includes(VTL) && type.includes(VTL)) {
		const vtl = executeVTL(expression, vtlBindings, logging) as ReactNode;
		if (features.includes(MD) && type.includes(MD) && typeof vtl === 'string')
			return <MdLabel expression={vtl} />;
		return vtl;
	}
	return expression;
}

export default executeExpression;
