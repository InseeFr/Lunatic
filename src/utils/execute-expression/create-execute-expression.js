import executeExpression from './execute-expression';
import getExpressionVariables from './get-expressions-variables';
import createMemoizer from './create-memoizer';
import createRefreshCalculated from './create-refresh-calculated';

function getVtlCompatibleValue(value) {
	if (value === undefined) {
		return null;
	}
	if (Array.isArray(value)) {
		return {
			dataStructure: { result: {} },
			dataPoints: {
				result: value,
			},
		};
	}

	return value;
}

function createBindings(variables) {
	return Object.entries(variables).reduce(function (
		bindings,
		[name, { values = {} }]
	) {
		const { PREVIOUS, COLLECTED, FORCED } = values;
		return { ...bindings, [name]: PREVIOUS || COLLECTED || FORCED || null };
	},
	{});
}

/**
 *
 * @param {*} variables
 * @returns
 */
function createExecuteExpression(source, features) {
	const { variables } = source;
	const bindings = createBindings(variables);
	const tokensMap = new Map();
	const collectedUpdated = new Map();
	const [memoize, getMemoizedValue] = createMemoizer();
	const [refreshCalculated, setToRefreshCalculated] = createRefreshCalculated({
		variables,
		execute,
		bindings,
	});

	/**
	 *
	 * @param {*} name
	 * @param {*} value
	 */
	function updateBindings(name, value) {
		// update des bindings
		// if (name in bindings) {
		// 	bindings[name] = value;
		// 	collectedUpdated.set(name, []);
		// }
		// // enrichissement des variables à rafraîchir
		// const { CalculatedLinked = [] } = variables[name];
		// CalculatedLinked.forEach(function (variable) {
		// 	const { name } = variable;
		// 	setToRefreshCalculated(name, variable);
		// });
	}

	function getVariablesAndCach(expression) {
		if (tokensMap.has(expression)) {
			return tokensMap.get(expression);
		}
		const tokens = getExpressionVariables(expression, variables);
		tokensMap.set(expression, tokens);
		return tokens;
	}

	/**/
	function collecteVariables(dependencies) {
		// if (Array.isArray(dependencies)) {
		// 	return dependencies.reduce(function (map, name) {
		// 		if (name in variables) {
		// 			const data = variables[name];
		// 			const { variable, type } = data;
		// 			if (!(name in map)) {
		// 				if (type === 'CALCULATED') {
		// 					const { expression } = variable;
		// 					const subDependencies = getVariablesAndCach(expression);
		// 					return {
		// 						...map,
		// 						[name]: { ...variable },
		// 						...collecteVariables(subDependencies),
		// 					};
		// 				}
		// 				return { ...map, [name]: { ...variable } };
		// 			}
		// 		} else {
		// 			throw new Error(`Unknow variable ${name}`);
		// 		}
		// 		return map;
		// 	}, {});
		// }
		// return {};
	}

	function resolveUseContext(name, { iteration }) {
		// const value = bindings[name];
		// if (iteration !== undefined && Array.isArray(value)) {
		// 	return value[iteration] || null;
		// }
		// return getVtlCompatibleValue(value);
	}

	function fillVariablesValues(map, { iteration }) {
		// return Object.entries(map).reduce(function (sub, [name, _]) {
		// 	return {
		// 		...sub,
		// 		[name]: resolveUseContext(name, { iteration }),
		// 	};
		// }, {});

		return map;
	}

	/*	*/

	function executeVTL(expression, args) {
		const { iterations, logging } = args;
		const bindingDependencies = getVariablesAndCach(expression);

		function loggingDefault(_, bindings, e) {
			if (process.env.NODE_ENV === 'development') {
				console.warn(`VTL error :  ${expression}`, { ...args }, { bindings });
				console.warn(e);
			}
		}

		const vtlBindings = refreshCalculated(
			fillVariablesValues(collecteVariables(bindingDependencies), {
				iterations,
			}),
			{ rootExpression: expression, iterations }
		);

		const memoized = getMemoizedValue(expression, vtlBindings);
		if (!memoized) {
			const result = executeExpression(
				vtlBindings,
				expression,
				features,
				logging || loggingDefault
			);
			memoize(expression, vtlBindings, result);

			return result;
		}
		return memoized;
	}

	function execute(expression, args) {
		if (typeof expression === 'string') {
			return expression;
		}
		if (typeof expression === 'object') {
			const { value, type } = expression;
			switch (type) {
				case 'VTL':
					return executeVTL(value, args);

				default:
					throw new Error(`Unknow type ${type}`);
			}
		}
	}

	return [execute, updateBindings];
}

export default createExecuteExpression;
