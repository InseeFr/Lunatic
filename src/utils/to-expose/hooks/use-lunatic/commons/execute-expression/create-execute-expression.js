import executeExpression from './execute-expression';

function getVtlCompatibleValue(value) {
	if (value === undefined) {
		return null;
	}
	if (Array.isArray(value)) {
		return {
			dataStructure: { result: {} },
			dataPoints: {
				result: value,
			},
		};
	}

	return value;
}

function createBindings(variables) {
	return Object.entries(variables).reduce(function (
		bindings,
		[name, { value }]
	) {
		return { ...bindings, [name]: value };
	},
	{});
}

/**
 *
 * @param {*} variables
 * @param {*} features
 * @returns
 */
function createExecuteExpression(variables, features) {
	// on aimerait map d'expression, avec les bindings
	const bindings = createBindings(variables);
	const expressionsMap = new Map();
	const toRefreshVariables = new Map(); // variables calculées dépendantes d'une variable modifiée.
	// à l'init, on y colle toutes les variables de calcul
	Object.values(variables).forEach(function ({ variable }) {
		const { variableType, name } = variable;
		if (variableType === 'CALCULATED') {
			toRefreshVariables.set(name, variable);
		}
	});
	/**
	 *
	 * @param {*} name
	 * @param {*} value
	 */
	function updateBindings(name, value) {
		// update des bindings
		if (name in bindings) {
			bindings[name] = value;
		}
		// enrichissement des variables à rafraîchir
		const { CalculatedLinked = [] } = variables[name];

		CalculatedLinked.forEach(function (variable) {
			const { name } = variable;
			toRefreshVariables.set(name, variable);
		});
	}

	/**/

	function collecteVariables(dependencies, variables) {
		if (Array.isArray(dependencies)) {
			return dependencies.reduce(function (map, name) {
				const data = variables[name];
				const { variable, type } = data;
				if (!(name in map)) {
					if (type === 'CALCULATED') {
						const { bindingDependencies: subDependencies } = variable;
						return {
							...map,
							[name]: { ...variable },
							...collecteVariables(subDependencies, variables),
						};
					}

					return { ...map, [name]: { ...variable } };
				}
				return map;
			}, {});
		}
		return {};
	}

	function resolveUseContext(name, { bindings, iteration }) {
		const value = bindings[name];
		if (iteration !== undefined && Array.isArray(value)) {
			return value[iteration];
		}
		return getVtlCompatibleValue(value);
	}

	function refreshCalculated(map, { variables, bindings, features }) {
		return Object.entries(map).reduce(function (sub, [name, current]) {
			const { variable, type } = variables[name];

			if (type === 'CALCULATED' && toRefreshVariables.has(name)) {
				const { expression } = variable;
				const value = executeExpression(map, expression, features);
				bindings[name] = value;
				toRefreshVariables.delete(name);

				return { ...sub, [name]: value };
			}
			return { ...sub, [name]: current };
		}, {});
	}

	function fillVariablesValues(map, { bindings, iteration }) {
		return Object.entries(map).reduce(function (sub, [name, _]) {
			return {
				...sub,
				[name]: resolveUseContext(name, { bindings, iteration }),
			};
		}, {});
	}

	/*	*/

	function directExecute(expression, args) {
		const { bindingDependencies, iteration } = args;

		function logging(_, e) {
			if (process.env.NODE_ENV === 'development') {
				console.warn(`VTL error :  ${expression}`, args);
				console.warn(e);
			}
		}

		const map = refreshCalculated(
			fillVariablesValues(collecteVariables(bindingDependencies, variables), {
				bindings,
				iteration,
				variables,
			}),
			{ variables, bindings, features }
		);
		const result = executeExpression(map, expression, features, logging);

		return result;
	}

	/**
	 *
	 * @param {*} expression
	 * @param {*} feature
	 * @param {*} param2
	 * @returns
	 */
	function execute(expression, args = {}) {
		const { bindingDependencies } = args;
		if (expressionsMap.has(expression)) {
			return expressionsMap.get(expression);
		}
		const value = directExecute(expression, args);
		if (!bindingDependencies) {
			expressionsMap.set(expression, value);
		}

		return value;
	}

	return [execute, updateBindings];
}

export default createExecuteExpression;
