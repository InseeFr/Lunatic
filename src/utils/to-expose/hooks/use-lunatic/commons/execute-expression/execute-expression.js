import { interpret } from '@inseefr/trevas';

function isDataSet(result) {
	return result && typeof result === 'object' && result.dataPoints;
}

function extractDataSetResult(dataSet) {
	const { dataPoints } = dataSet;
	if (dataPoints) {
		const { result } = dataPoints;
		return result;
	}
	return undefined;
}

function executeVtl(expression, vtlBindings) {
	const result = interpret(expression, vtlBindings);
	if (isDataSet(result)) {
		return extractDataSetResult(result);
	} else if (typeof result === 'object') {
		return expression;
	}

	return result;
}

function loggingDefault(expression, bindings, e) {
	if (process.env.NODE_ENV === 'development') {
		console.warn(`VTL error :  ${expression}`, bindings);
		console.warn(e);
	}
}

function executeExpression(
	vtlBindings,
	expression,
	features /* VTL, MD */,
	logging = loggingDefault
) {
	if (expression) {
		try {
			if (features.includes('VTL')) {
				const result = executeVtl(expression, vtlBindings);

				// console.log({ expression, result, vtlBindings });
				return result;
			}
			return expression;
		} catch (e) {
			// expression en erreur ou simple chaîne de caractère
			logging(expression, vtlBindings, e);
			return expression;
		}
		// TODO MD only for labels, not for filtering
	}
	return '';
}

export default executeExpression;
