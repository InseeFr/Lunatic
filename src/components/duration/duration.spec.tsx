import React from 'react';
import { render } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import Duration from './duration';
import { ExpressionLogger } from '../../use-lunatic/commons/execute-expression/create-execute-expression';

describe('Duration component', () => {
	const handleChange = vi.fn();

	it('should render with hours and minutes inputs when format is "PTnHnM"', () => {
		const { getByLabelText } = render(
			<Duration
				format="PTnHnM"
				handleChange={handleChange}
				id={''}
				value={null}
				executeExpression={function <T extends unknown = unknown>(
					expression: unknown,
					args?:
						| {
								iteration?: number | undefined;
								linksIterations?: number[] | undefined;
								logging?: ExpressionLogger | undefined;
								bindingDependencies?: string[] | undefined;
						  }
						| undefined
				): T {
					throw new Error('Function not implemented.');
				}}
				response={{
					name: '',
				}}
			/>
		);

		const hoursInput = getByLabelText('Heures:');
		const minutesInput = getByLabelText('Minutes:');

		expect(hoursInput).toBeInTheDocument();
		expect(minutesInput).toBeInTheDocument();
	});

	it('should render with years and months inputs when format is "PnYnM"', () => {
		const { getByLabelText } = render(
			<Duration
				format="PnYnM"
				handleChange={handleChange}
				id={''}
				value={null}
				executeExpression={function <T extends unknown = unknown>(
					expression: unknown,
					args?:
						| {
								iteration?: number | undefined;
								linksIterations?: number[] | undefined;
								logging?: ExpressionLogger | undefined;
								bindingDependencies?: string[] | undefined;
						  }
						| undefined
				): T {
					throw new Error('Function not implemented.');
				}}
				response={{
					name: '',
				}}
			/>
		);

		const yearsInput = getByLabelText('Ann√©es:');
		const monthsInput = getByLabelText('Mois:');

		expect(yearsInput).toBeInTheDocument();
		expect(monthsInput).toBeInTheDocument();
	});
});
