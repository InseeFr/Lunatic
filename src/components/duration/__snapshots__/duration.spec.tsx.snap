// Vitest Snapshot v1

exports[`Duration > renders without crashing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="field-container"
      >
        <div
          class="field"
        >
          <div
            class="container"
          >
            <div />
            <div
              class="field-group"
            >
              <div
                class="input-field-group"
              >
                <label
                  class="input-label"
                  for="hoursInput"
                >
                  Heures:
                </label>
                <input
                  class="input-field"
                  id="hoursInput"
                  max="23"
                  min="0"
                  type="number"
                  value=""
                />
              </div>
              <div
                class="input-field-group"
              >
                <label
                  class="input-label"
                  for="minutesInput"
                >
                  Minutes:
                </label>
                <input
                  class="input-field"
                  id="minutesInput"
                  max="59"
                  min="0"
                  type="number"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="field-container"
    >
      <div
        class="field"
      >
        <div
          class="container"
        >
          <div />
          <div
            class="field-group"
          >
            <div
              class="input-field-group"
            >
              <label
                class="input-label"
                for="hoursInput"
              >
                Heures:
              </label>
              <input
                class="input-field"
                id="hoursInput"
                max="23"
                min="0"
                type="number"
                value=""
              />
            </div>
            <div
              class="input-field-group"
            >
              <label
                class="input-label"
                for="minutesInput"
              >
                Minutes:
              </label>
              <input
                class="input-field"
                id="minutesInput"
                max="59"
                min="0"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
